### ISSUE TITLE
Adding postal/zip codes

### ISSUE THREAD

**Issue by dagnelies** [OSMNames/OSMNames#191]
Hi,

It would be nice to add postal/zip codes to the entries. IMHO it's one of the most important missing features to really have complete addresses.

I noticed that there were a couple of posts in the past pointing out that it is not trivial/reliable. However, even if imperfect/incomplete, it would still be a great addition.

**Comment by sfkeller** [comment]
Hi,
Postal addresses could be extraced from OSM, yes. 
But postal addresses are really a different thing than addresses: 
They are controlled by an entity (Post) which is often private, they are fast changing or even inexistant in some countries.
So according to my long experience, postal/zip codes don't havy any priority here.
If you want to have postal/zip codes I'd recommend e.g. https://openaddresses.io/ which is a bucket for such data.
I'm closing this but won't discourage you to still keep enhancing OSMNames.

**Comment by dagnelies** [comment]
> They are controlled by an entity (Post) which is often private, they are fast changing or even inexistant in some countries.

I am suprised by this statement since my perception is quite the opposite. I thought postal/zip codes to be something well defined and rarely changing over time.

https://gis.stackexchange.com/questions/111470/average-frequency-of-zip-code-change-for-a-given-address-over-the-course-of-year

According to his estimate 0.01% of streets change their ZIP codes within a year. I assume european countries are similar, if not even more stable.

Even if 0.01% of the data is "erroneous", isn't it still "good enough"/useful? ;)

Sadly https://openaddresses.io/ has too much missing data/areas to be useful. Actually, if OSMNames had postal/zip codes, it would be the perfect replacement for openaddresses.

**Comment by masbaehr** [comment]
I was also thinking to use osmnames, for facilitating manual address input by users of my application , however i also quickly missed the ZIP/CAP. Does anyone know if there is a way to support those too? Even if it meant to do some lookup via a different service ? E.g. openaddresses. Adding zip to osmnames should still not cause any harm imho. If the concern is that they are "fast changing" - which in certain countries - like germany, each street has a ZIP which hardly ever changes - There are other things like "class" ("highway") or "type" (residential) - which change even more often than the zip. 

Just for understanding - Does OSMNames not use openstreetmaps data? Because there i see the Postcode.?

**Comment by philippks** [comment]
@masbaehr OSMNames uses OpenStreetMap data and therefore it is possible to add the postcode. It would be a new step in the [prepare_data workflow](https://github.com/OSMNames/OSMNames/blob/master/osmnames/prepare_data/prepare_data.py).

@sfkeller do you still think it makes no sense to add this feature to osmnames or is it just a matter of priority?

**Comment by klokan** [comment]
I believe it make sense to do that.


If somebody is willing to work on this and make a good Pull Request, then it can be merged...

**Comment by dagnelies** [comment]
Well, then let us re-open that ticket ;)

I remember I looked at it in the past, and I located the places to enhance. However, reproducing the environment to run the whole thing to test it proved to be far from trivial. If I do find some time, I will look into it again.

**Comment by philippks** [comment]
Great. @dagnelies let me know if you need any help. For testing, you can run it for a small country first. I can run it on the full planet as soon as it is ready.

**Comment by sfkeller** [comment]
> @sfkeller do you still think it makes no sense to add this feature to osmnames or is it just a matter of priority?

I agree with @klokan not to no throw away postal/zip codes in general and accepting good Pull Requests (which are not too much code). But with a great "BUT": Don't expect a good worldwide solution in the first place! 

One has to be aware that postal/zip codes are organized very, very differently in each country - besides being not yet fully mapped in OSM. In some countries they are even either useless, or not accessible (private postal companies), or changing fast!

In Switzerland for example there's a new "Amtliches Ortschaftenverzeichnis mit Postleitzahl und Perimeter" every month. In this case one would ignore postal/zip codes from OSM and instead calculate an overlay/ the official polygons in a post-process. So, what's needed to fulfill the expectations, are adaptions country by country.

Perhaps we can reach this functionality in a step-wise approach
1. Add it worldwide - and evaluating the results (works, doesn't-work, could-work-with-local-adaptions) 
2. Publish a list of countries where postal/zip make sense (whereas in other country one would better leave them out and e.g. ignore them even in search in order e.g. not to confuse with house numbers)
3. Add solutions for selected countries step-by-step.

So I'm glad if @dagnelies makes an digestible and well written Pull Request. Is this trying to be a generic solution - or one geared e.g. to France?

**Comment by dagnelies** [comment]
...duh ...I'm kind of swamped with work lately ...I don't think I'll find the time to do it for a while. Sorry if I had your hopes up 😞

**Comment by sfkeller** [comment]
Thx for your info. Hope you still keep involved whenever you have time.

**Comment by dagnelies** [comment]
Well, after a long time, I wanted to get back to this issue, but I'm facing difficulties.

If I understand correctly, determining the city, state etc. for a street is based on the `get_parent_info` and on a strict hierarchy: a city is in a county which is in a state, etc. However, postal codes have distinct boundaries. There are some cities divided into several postal codes, and postal codes covering several smaller cities. There are also streets crossing two postal codes. So it's not straightforward.

**Comment by sfkeller** [comment]
> However, postal codes have distinct boundaries. 

Yes, postal codes are mostly different from administration boundaries. 

But no, postal codes have no "boundaries" - at least not in Switzerland and other countries, where postal codes are defined purely by optimized delivery routes and delivery centers.

**Comment by dagnelies** [comment]
That's surprising, in germany at least they have clear boundaries.

There is also a specific boundary type in the OSM for it:

https://wiki.openstreetmap.org/wiki/Tag:boundary%3Dpostal_code

**Comment by dagnelies** [comment]
I just picker the Luxembourg OSM to check the raw data as an example ...and ...it's quite odd/inconsistent. 

There are:

- very few "postal_code" boundaries (totally incomplete)
- some "postal_code" tags attached directly to the roads (very incomplete)
- huge amount of "addr:postcode" attached to houses/buildings (complete)

**Comment by sfkeller** [comment]
> That's surprising, in germany at least they have clear boundaries.

One can derive and buy boundaries for Swiss postal/zip codes too. But they are concave hulls derived from point info.
